age/keysource.go

```diff
@@ -326,11 +328,88 @@ func (key *MasterKey) loadIdentities() (ParsedIdentities, error) {
 	}
 
 	for n, r := range readers {
-		ids, err := age.ParseIdentities(r)
-		if err != nil {
-			return nil, fmt.Errorf("failed to parse '%s' age identities: %w", n, err)
+
+		b := bufio.NewReader(r)
+		p, _ := b.Peek(14) // length of "age-encryption" and "-----BEGIN AGE"
+		peeked := string(p)
+
+		switch {
+		// An age encrypted file, plain or armored.
+		case peeked == "age-encryption" || peeked == "-----BEGIN AGE":
+			var r io.Reader = b
+			if peeked == "-----BEGIN AGE" {
+				r = armor.NewReader(r)
+			}
+			const privateKeySizeLimit = 1 << 24 // 16 MiB
+			contents, err := io.ReadAll(io.LimitReader(r, privateKeySizeLimit))
+			if err != nil {
+				return nil, fmt.Errorf("failed to read '%s': %w", n, err)
+			}
+			if len(contents) == privateKeySizeLimit {
+				return nil, fmt.Errorf("failed to read '%s': file too long", n)
+			}
+			IncorrectPassphrase := func() {
+				conn, err := gpgagent.NewConn()
+				if err != nil {
+					return
+				}
+				defer func(conn *gpgagent.Conn) {
+					if err := conn.Close(); err != nil {
+						log.Errorf("failed to close connection with gpg-agent: %s", err)
+					}
+				}(conn)
+				err = conn.RemoveFromCache(n)
+				if err != nil {
+					log.Warnf("gpg-agent remove cache request errored: %s", err)
+					return
+				}
+			}
+			ids := []age.Identity{&EncryptedIdentity{
+				Contents: contents,
+				Passphrase: func() (string, error) {
+					conn, err := gpgagent.NewConn()
+					if err != nil {
+						passphrase, err := readPassphrase("Enter passphrase for identity " + n + ":")
+						if err != nil {
+							return "", err
+						}
+						return string(passphrase), nil
+					}
+					defer func(conn *gpgagent.Conn) {
+						if err := conn.Close(); err != nil {
+							log.Errorf("failed to close connection with gpg-agent: %s", err)
+						}
+					}(conn)
+
+					req := gpgagent.PassphraseRequest{
+						// TODO is the cachekey good enough?
+						CacheKey: n,
+						Prompt:   "Passphrase",
+						Desc:     fmt.Sprintf("Enter passphrase for identity '%s':", n),
+					}
+					pass, err := conn.GetPassphrase(&req)
+					if err != nil {
+						return "", fmt.Errorf("gpg-agent passphrase request errored: %s", err)
+					}
+					//make sure that we won't store empty pass
+					if len(pass) == 0 {
+						IncorrectPassphrase()
+					}
+					return pass, nil
+				},
+				IncorrectPassphrase: IncorrectPassphrase,
+				NoMatchWarning: func() {
+					log.Warnf("encrypted identity '%s' didn't match file's recipients", n)
+				},
+			}}
+			identities = append(identities, ids...)
+		default:
+			ids, err := age.ParseIdentities(b)
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse '%s' age identities: %w", n, err)
+			}
+			identities = append(identities, ids...)
```

Comment:
Since this code is also heavily copied from https://github.com/FiloSottile/age/blob/101cc8676386b0503571a929a88618cae2f0b1cd/cmd/age/parse.go, it's better to move it to a separate file (also `encrypted_keys.go`?) Also moving this part to its own function (that can be moved to that file) will improve the code anyway, the function is getting quite long.

Reply:
done